toolbox/deps_update/main.go

```diff
@@ -0,0 +1,230 @@
+	if err != nil {
+		return err
+	}
+	mixerSHA, err := githubClnt.getHeadCommitSHA("mixer", "stable")
+	if err != nil {
+		return err
+	}
+	AuthSHA, err := githubClnt.getHeadCommitSHA("auth", "stable")
+	if err != nil {
+		return err
+	}
+	caHub := "docker.io/istio"
+	istioctlURL := fmt.Sprintf(
+		"https://storage.googleapis.com/istio-artifacts/pilot/%s/artifacts/istioctl", pilotSHA)
+	hub := "gcr.io/istio-testing"
+	cmd := fmt.Sprintf("./install/updateVersion.sh -p %s,%s -x %s,%s -i %s -c %s,%s",
+		hub, pilotSHA, hub, mixerSHA, istioctlURL, caHub, AuthSHA)
+	_, err = util.Shell(cmd)
+	return err
+}
+
+// Update the commit SHA reference in the dependency file of dep
+func updateDeps(repo string, deps []dependency) error {
+	if repo == "istio" {
+		if err := updateIstioDeps(); err != nil {
+			return err
+		}
+	} else {
+		for _, dep := range deps {
+			if err := updateDepFile(dep); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// Delete the local git repo just cloned
+func cleanUp(repo string) error {
+	if err := os.Chdir(".."); err != nil {
+		return err
+	}
+	return os.RemoveAll(repo)
+}
+
+// Update the given repository so that it uses the latest dependency references
+// push new branch to remote, create pull request on master,
+// which is auto-merged after presumbit
+func updateDependenciesOf(repo string) error {
+	if err := os.RemoveAll(repo); err != nil {
+		return err
+	}
+	if _, err := util.Shell("git clone " + githubClnt.remote(repo)); err != nil {
+		return err
+	}
+	if err := os.Chdir(repo); err != nil {
+		return err
+	}
+	defer func() {
+		if err := cleanUp(repo); err != nil {
+			log.Fatalf("Error during clean up: %v\n", err)
+		}
+	}()
+
+	if _, err := util.Shell("git checkout deps"); err != nil {
+		return err
+	}
+
+	deps, err := deserializeDeps(istioDepsFile)
+	if err != nil {
+		return err
+	}
+	depVersions, err := fingerPrintAndUpdateDepSHA(repo, &deps)
+	if err != nil {
+		return err
+	}
+	branch := "autoUpdateDeps_" + depVersions
+	exists, err := githubClnt.existBranch(repo, branch)
+	if err != nil {
+		return err
+	}
+	if exists {
+		log.Printf("Branch already exists")
+	}
+	// if branch exists, stop here and do not create another PR of identical delta
+	if err = githubClnt.closeFailedPullRequests(repo); exists || err != nil {
+		return err
+	}
+	if _, err := util.Shell("git checkout -b " + branch); err != nil {
+		return err
+	}
+	if err := updateDeps(repo, deps); err != nil {
+		return err
+	}
+	if err := serializeDeps(istioDepsFile, &deps); err != nil {
+		return err
+	}
+	if _, err := util.Shell("git add *"); err != nil {
+		return err
+	}
+	if _, err := util.Shell("git commit -m Update_Dependencies"); err != nil {
+		return err
+	}
+	if _, err := util.Shell("git push --set-upstream origin " + branch); err != nil {
+		return err
+	}
+	if err := githubClnt.createPullRequest(branch, repo); err != nil {
+		return err
+	}
+	return nil
+}
+
+func main() {
```

Comment:
I suppose it is a periodic job, then why not use goroutine and wake up every midnight